File: C++/binary_search/find_peak_element.cpp
/*
Problem: Find Peak Element (LeetCode 162)
Difficulty: Medium
Topic: Binary Search
Description: A peak element is an element that is strictly greater than its neighbors.
Given a 0-indexed integer array nums, find a peak element, and return its index. If the array contains multiple peaks, return the index to any of the peaks.
You may imagine that nums[-1] = nums[n] = -∞. In other words, an element is always considered to be strictly greater than a neighbor that is outside the array.
You must write an algorithm that runs in O(log n) time.

Time Complexity: O(log n)
Space Complexity: O(1)
*/

/*
Algorithm intuition:
- Use binary search by checking the slope at mid point
- If nums[mid] < nums[mid+1], peak must be on the right side (go right)
- If nums[mid] > nums[mid+1], peak must be on the left side or mid itself (go left)
- This works because we're guaranteed to find a peak due to the -∞ boundary conditions
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int left = 0, right = nums.size() - 1;
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            
            // If mid element is smaller than next element,
            // peak must be on the right side
            if (nums[mid] < nums[mid + 1]) {
                left = mid + 1;
            } 
            // If mid element is greater than or equal to next element,
            // peak must be on the left side (including mid)
            else {
                right = mid;
            }
        }
        
        return left; // left == right at this point
    }
};

// Test function
int main() {
    Solution solution;
    
    // Test case 1: [1,2,3,1]
    vector<int> nums1 = {1, 2, 3, 1};
    int result1 = solution.findPeakElement(nums1);
    cout << "Test 1: " << result1 << " (Expected: 2, value: " << nums1[result1] << ")" << endl;
    
    // Test case 2: [1,2,1,3,5,6,4]
    vector<int> nums2 = {1, 2, 1, 3, 5, 6, 4};
    int result2 = solution.findPeakElement(nums2);
    cout << "Test 2: " << result2 << " (Expected: 1 or 5, value: " << nums2[result2] << ")" << endl;
    
    // Test case 3: [1]
    vector<int> nums3 = {1};
    int result3 = solution.findPeakElement(nums3);
    cout << "Test 3: " << result3 << " (Expected: 0, value: " << nums3[result3] << ")" << endl;
    
    // Test case 4: [1,2]
    vector<int> nums4 = {1, 2};
    int result4 = solution.findPeakElement(nums4);
    cout << "Test 4: " << result4 << " (Expected: 1, value: " << nums4[result4] << ")" << endl;
    
    return 0;
}

File: Java/searching/FindPeakElement.java
/**
 * This class finds a peak element in an array where a peak element is defined as
 * an element that is strictly greater than its neighbors.
 * Note: For edge elements, we only need to compare with the inner neighbor.
 * 
 * Example:
 * Input: nums = [1,2,3,1]
 * Output: 2 (3 is a peak element because it's greater than its neighbors)
 * 
 * Time Complexity: O(log n) - Binary Search approach
 * Space Complexity: O(1) - Constant extra space
 */
public class FindPeakElement {
    
    /**
     * Finds a peak element in the array using binary search.
     * The function assumes that nums[i] != nums[i + 1] for all valid i.
     * 
     * @param nums the input array to find peak element in
     * @return the index of any peak element in the array
     */
    public int findPeakElement(int[] nums) {
        // Edge case: single element array
        if (nums.length == 1) {
            return 0;
        }
        
        int low = 0;
        int high = nums.length - 1;
        
        while (low < high) {
            int mid = low + (high - low) / 2;  // Prevents integer overflow
            
            // If mid element is greater than next element,
            // a peak exists in the left half (including mid)
            if (nums[mid] > nums[mid + 1]) {
                high = mid;
            }
            // If mid element is less than next element,
            // a peak exists in the right half
            else {
                low = mid + 1;
            }
        }
        
        // At this point, low == high and points to the peak element
        return low;
    }

    /**
     * Main method to test the peak finding functionality with example cases.
     * 
     * @param args command line arguments (not used)
     */
    public static void main(String[] args) {
        FindPeakElement solution = new FindPeakElement();
        
        // Test case 1: Standard case with peak in middle
        int[] nums1 = {1, 2, 3, 1};
        System.out.println("Test case 1:");
        System.out.println("Array: [1, 2, 3, 1]");
        int peak1 = solution.findPeakElement(nums1);
        System.out.println("Peak element index: " + peak1);
        System.out.println("Peak element value: " + nums1[peak1]);
        System.out.println();
        
        // Test case 2: Multiple peaks (returns any peak)
        int[] nums2 = {1, 2, 1, 3, 5, 6, 4};
        System.out.println("Test case 2:");
        System.out.println("Array: [1, 2, 1, 3, 5, 6, 4]");
        int peak2 = solution.findPeakElement(nums2);
        System.out.println("Peak element index: " + peak2);
        System.out.println("Peak element value: " + nums2[peak2]);
        System.out.println();
        
        // Test case 3: Peak at the end
        int[] nums3 = {1, 2, 3, 4};
        System.out.println("Test case 3:");
        System.out.println("Array: [1, 2, 3, 4]");
        int peak3 = solution.findPeakElement(nums3);
        System.out.println("Peak element index: " + peak3);
        System.out.println("Peak element value: " + nums3[peak3]);
        System.out.println();
        
        // Test case 4: Single element array
        int[] nums4 = {1};
        System.out.println("Test case 4:");
        System.out.println("Array: [1]");
        int peak4 = solution.findPeakElement(nums4);
        System.out.println("Peak element index: " + peak4);
        System.out.println("Peak element value: " + nums4[peak4]);
    }
}